In this project we are going to do EDA of Telecome data and we will visualise the data by using python and python libraries.

The objective of this project is to identify the factor of customer charm. 

In [24]:
#Import required libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [25]:
#read the data from pc and make dataframe
df = pd.read_csv(r'C:\Users\HP\Desktop\Data-LB\New folder\project2.csv')

In [26]:
#read top 10 row of data
df.head(10)

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   
5  9305-CDSKC  Female              0      No         No       8          Yes   
6  1452-KIOVK    Male              0      No        Yes      22          Yes   
7  6713-OKOMC  Female              0      No         No      10           No   
8  7892-POOKP  Female              0     Yes         No      28          Yes   
9  6388-TABGU    Male              0      No        Yes      62          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   
5               Yes     Fiber optic             No  ...              Yes   
6               Yes     Fiber optic             No  ...               No   
7  No phone service             DSL            Yes  ...               No   
8               Yes     Fiber optic             No  ...              Yes   
9                No             DSL            Yes  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   
5          No         Yes             Yes  Month-to-month              Yes   
6          No         Yes              No  Month-to-month              Yes   
7          No          No              No  Month-to-month               No   
8         Yes         Yes             Yes  Month-to-month              Yes   
9          No          No              No        One year               No   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  
5           Electronic check          99.65         820.5   Yes  
6    Credit card (automatic)          89.10        1949.4    No  
7               Mailed check          29.75         301.9    No  
8           Electronic check         104.80       3046.05   Yes  
9  Bank transfer (automatic)          56.15       3487.95    No  

[10 rows x 21 columns]

The features and the target can be identify by the visual inspection of above data frame.

The target variable: The column named 'churn' which contain whether customer is churned out or not(Yes/No)
Apart from the target, they can be further divide into two categories.

* Customer Demography features
* customer services offered

from data we can see that all column's name not consistance in capital letter or small letter.

In [27]:
#convert all column's name in small letters
df.columns = [col.lower() for col in df]

In [28]:
#check the column name by printing top 5 row of data
df.head()

customerid  gender  seniorcitizen partner dependents  tenure phoneservice  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      multiplelines internetservice onlinesecurity  ... deviceprotection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  techsupport streamingtv streamingmovies        contract paperlessbilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               paymentmethod monthlycharges  totalcharges churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

Now all column's names are consistance. all names are in small letters.

In [29]:
#get the summary of data, data type, count
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerid        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   seniorcitizen     7043 non-null   int64  
 3   partner           7043 non-null   object 
 4   dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   phoneservice      7043 non-null   object 
 7   multiplelines     7043 non-null   object 
 8   internetservice   7043 non-null   object 
 9   onlinesecurity    7043 non-null   object 
 10  onlinebackup      7043 non-null   object 
 11  deviceprotection  7043 non-null   object 
 12  techsupport       7043 non-null   object 
 13  streamingtv       7043 non-null   object 
 14  streamingmovies   7043 non-null   object 
 15  contract          7043 non-null   object 
 16  paperlessbilling  7043 non-null   object 


From above summary we can say that there are 21 columns and 7043 rows. And there is no missing value.

In [30]:
#Check Duplicate row in data
df.duplicated().sum()

0

There is no duplicate value in data.

In [31]:
#decription of data
df.describe()

seniorcitizen       tenure  monthlycharges
count    7043.000000  7043.000000     7043.000000
mean        0.162147    32.371149       64.761692
std         0.368612    24.559481       30.090047
min         0.000000     0.000000       18.250000
25%         0.000000     9.000000       35.500000
50%         0.000000    29.000000       70.350000
75%         0.000000    55.000000       89.850000
max         1.000000    72.000000      118.750000

From above description we got the statistics for Senior citizen, tenure and monthly charges.

In [32]:
#Find the no. of unique values in each column

for i in df.columns:
    print(f"uniques {i}'s count: {df[i].nunique()}")
    
    #print unique values
    print(f"{df[i].unique()}\n")

uniques customerid's count: 7043
['7590-VHVEG' '5575-GNVDE' '3668-QPYBK' ... '4801-JZAZL' '8361-LTMKD'
 '3186-AJIEK']

uniques gender's count: 2
['Female' 'Male']

uniques seniorcitizen's count: 2
[0 1]

uniques partner's count: 2
['Yes' 'No']

uniques dependents's count: 2
['No' 'Yes']

uniques tenure's count: 73
[ 1 34  2 45  8 22 10 28 62 13 16 58 49 25 69 52 71 21 12 30 47 72 17 27
  5 46 11 70 63 43 15 60 18 66  9  3 31 50 64 56  7 42 35 48 29 65 38 68
 32 55 37 36 41  6  4 33 67 23 57 61 14 20 53 40 59 24 44 19 54 51 26  0
 39]

uniques phoneservice's count: 2
['No' 'Yes']

uniques multiplelines's count: 3
['No phone service' 'No' 'Yes']

uniques internetservice's count: 3
['DSL' 'Fiber optic' 'No']

uniques onlinesecurity's count: 3
['No' 'Yes' 'No internet service']

uniques onlinebackup's count: 3
['Yes' 'No' 'No internet service']

uniques deviceprotection's count: 3
['No' 'Yes' 'No internet service']

uniques techsupport's count: 3
['No' 'Yes' 'No internet service']

uniques

From above data we can say that in some columns Nominal data is given as data type object. Now we will convert all these data as integer.

In [33]:
#convert the data as 1 for yes and 0 for No
df['partner'] =df['partner'].map({'Yes':1,'No':0})
df['dependents']=df['dependents'].map({'Yes':1,'No':0})
df['paperlessbilling']=df['paperlessbilling'].map({'Yes':1,'No':0})
df['churn']=df['churn'].map({'Yes':1,'No':0})

In [34]:
#check summary
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerid        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   seniorcitizen     7043 non-null   int64  
 3   partner           7043 non-null   int64  
 4   dependents        7043 non-null   int64  
 5   tenure            7043 non-null   int64  
 6   phoneservice      7043 non-null   object 
 7   multiplelines     7043 non-null   object 
 8   internetservice   7043 non-null   object 
 9   onlinesecurity    7043 non-null   object 
 10  onlinebackup      7043 non-null   object 
 11  deviceprotection  7043 non-null   object 
 12  techsupport       7043 non-null   object 
 13  streamingtv       7043 non-null   object 
 14  streamingmovies   7043 non-null   object 
 15  contract          7043 non-null   object 
 16  paperlessbilling  7043 non-null   int64  


Now total integer columns = 5 and float = 2.

By the visual inspection we saw that totalcharges have numerical values but in summary it is showing data type of totalcharges as object. Hence we will convert the data type of this column.

In [35]:
#convert data type from object to float of totalcharges
df['totalcharges'] = pd.to_numeric(df['totalcharges'],errors = 'coerce')

In [36]:
#Again check summary
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerid        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   seniorcitizen     7043 non-null   int64  
 3   partner           7043 non-null   int64  
 4   dependents        7043 non-null   int64  
 5   tenure            7043 non-null   int64  
 6   phoneservice      7043 non-null   object 
 7   multiplelines     7043 non-null   object 
 8   internetservice   7043 non-null   object 
 9   onlinesecurity    7043 non-null   object 
 10  onlinebackup      7043 non-null   object 
 11  deviceprotection  7043 non-null   object 
 12  techsupport       7043 non-null   object 
 13  streamingtv       7043 non-null   object 
 14  streamingmovies   7043 non-null   object 
 15  contract          7043 non-null   object 
 16  paperlessbilling  7043 non-null   int64  


Now data type of totalcharges has been changed. But there are only 7032 values.

In [37]:
df.isnull().sum()

customerid           0
gender               0
seniorcitizen        0
partner              0
dependents           0
tenure               0
phoneservice         0
multiplelines        0
internetservice      0
onlinesecurity       0
onlinebackup         0
deviceprotection     0
techsupport          0
streamingtv          0
streamingmovies      0
contract             0
paperlessbilling     0
paymentmethod        0
monthlycharges       0
totalcharges        11
churn                0
dtype: int64

There are only 11 null values present in data set. 

In [38]:
#Delete the null values

In [39]:
df.dropna(inplace = True)

In [40]:
df

customerid  gender  seniorcitizen  partner  dependents  tenure  \
0     7590-VHVEG  Female              0        1           0       1   
1     5575-GNVDE    Male              0        0           0      34   
2     3668-QPYBK    Male              0        0           0       2   
3     7795-CFOCW    Male              0        0           0      45   
4     9237-HQITU  Female              0        0           0       2   
...          ...     ...            ...      ...         ...     ...   
7038  6840-RESVB    Male              0        1           1      24   
7039  2234-XADUH  Female              0        1           1      72   
7040  4801-JZAZL  Female              0        1           1      11   
7041  8361-LTMKD    Male              1        1           0       4   
7042  3186-AJIEK    Male              0        0           0      66   

     phoneservice     multiplelines internetservice onlinesecurity  ...  \
0              No  No phone service             DSL             No  ...   
1             Yes                No             DSL            Yes  ...   
2             Yes                No             DSL            Yes  ...   
3              No  No phone service             DSL            Yes  ...   
4             Yes                No     Fiber optic             No  ...   
...           ...               ...             ...            ...  ...   
7038          Yes               Yes             DSL            Yes  ...   
7039          Yes               Yes     Fiber optic             No  ...   
7040           No  No phone service             DSL            Yes  ...   
7041          Yes               Yes     Fiber optic             No  ...   
7042          Yes                No     Fiber optic            Yes  ...   

     deviceprotection techsupport streamingtv streamingmovies        contract  \
0                  No          No          No              No  Month-to-month   
1                 Yes          No          No              No        One year   
2                  No          No          No              No  Month-to-month   
3                 Yes         Yes          No              No        One year   
4                  No          No          No              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7038              Yes         Yes         Yes             Yes        One year   
7039              Yes          No         Yes             Yes        One year   
7040               No          No          No              No  Month-to-month   
7041               No          No          No              No  Month-to-month   
7042              Yes         Yes         Yes             Yes        Two year   

     paperlessbilling              paymentmethod monthlycharges  totalcharges  \
0                   1           Electronic check          29.85         29.85   
1                   0               Mailed check          56.95       1889.50   
2                   1               Mailed check          53.85        108.15   
3                   0  Bank transfer (automatic)          42.30       1840.75   
4                   1           Electronic check          70.70        151.65   
...               ...                        ...            ...           ...   
7038                1               Mailed check          84.80       1990.50   
7039                1    Credit card (automatic)         103.20       7362.90   
7040                1           Electronic check          29.60        346.45   
7041                1               Mailed check          74.40        306.60   
7042                1  Bank transfer (automatic)         105.65       6844.50   

      churn  
0         0  
1         0  
2         1  
3         0  
4         1  
...     ...  
7038      0  
7039      0  
7040      0  
7041      1  
7042      0  

[7032 rows x 21 columns]

After removing null values, Now total number of row is equal to 7032.

Finally we have been cleaned the data. We will do EDA of data.

In [41]:
#print no. of churn out and not chrun out
df.groupby(df['churn']).count()

customerid  gender  seniorcitizen  partner  dependents  tenure  \
churn                                                                   
0            5163    5163           5163     5163        5163    5163   
1            1869    1869           1869     1869        1869    1869   

       phoneservice  multiplelines  internetservice  onlinesecurity  \
churn                                                                 
0              5163           5163             5163            5163   
1              1869           1869             1869            1869   

       onlinebackup  deviceprotection  techsupport  streamingtv  \
churn                                                             
0              5163              5163         5163         5163   
1              1869              1869         1869         1869   

       streamingmovies  contract  paperlessbilling  paymentmethod  \
churn                                                               
0                 5163      5163              5163           5163   
1                 1869      1869              1869           1869   

       monthlycharges  totalcharges  
churn                                
0                5163          5163  
1                1869          1869

Now we can see, total number of chunred out is equal to 1869 and number people who have not churned out = 5163

From above data we are not able to see that at which factor churn is depending. So we do fruther analysis.

In [46]:
#Make a pivot table
piv_tab = pd.pivot_table(df,index=['churn','gender','partner'],values = 'customerid',aggfunc = 'count')
piv_tab

customerid
churn gender partner            
0     Female 0              1180
             1              1364
      Male   0              1259
             1              1360
1     Female 0               620
             1               319
      Male   0               580
             1               350

From this data, it is not eassy to say what is the factor for churn. But we can say The Females who have partner, there is less probability of churn.

In [48]:
#Correlation between churn and other columns
df.corr()

seniorcitizen   partner  dependents    tenure  \
seniorcitizen          1.000000  0.016957   -0.210550  0.015683   
partner                0.016957  1.000000    0.452269  0.381912   
dependents            -0.210550  0.452269    1.000000  0.163386   
tenure                 0.015683  0.381912    0.163386  1.000000   
paperlessbilling       0.156258 -0.013957   -0.110131  0.004823   
monthlycharges         0.219874  0.097825   -0.112343  0.246862   
totalcharges           0.102411  0.319072    0.064653  0.825880   
churn                  0.150541 -0.149982   -0.163128 -0.354049   

                  paperlessbilling  monthlycharges  totalcharges     churn  
seniorcitizen             0.156258        0.219874      0.102411  0.150541  
partner                  -0.013957        0.097825      0.319072 -0.149982  
dependents               -0.110131       -0.112343      0.064653 -0.163128  
tenure                    0.004823        0.246862      0.825880 -0.354049  
paperlessbilling          1.000000        0.351930      0.157830  0.191454  
monthlycharges            0.351930        1.000000      0.651065  0.192858  
totalcharges              0.157830        0.651065      1.000000 -0.199484  
churn                     0.191454        0.192858     -0.199484  1.000000

From this correlation data we can say, there is no strong correlartion of churn with any other factor. According to this data strongest negative correlation of churn with tenure that is -0.354.

# Now we will go visualisation of data.